src/main/kotlin/com/exactpro/th2/codec/DecodeProcessor.kt

```diff
@@ -41,45 +44,48 @@ class DecodeProcessor(
                 continue
             }
 
-            val parentEventId = messageGroup.allParentEventIds
-
             if (messageGroup.messagesList.none(AnyMessage::hasRawMessage)) {
-                parentEventId.onErrorEvent("Message group has no parsed messages in it", messageGroup.messageIds)
+                logger.debug { "Message group has no raw messages in it" }
+                messageBatch.addGroups(messageGroup)
                 continue
             }
 
-            if (!messageGroup.isDecodable()) {
-                val info = "No messages of $protocol protocol or mixed empty and non-empty protocols are present"
-                parentEventId.onErrorEvent(info, messageGroup.messageIds)
-                messageBatch.addGroups(messageGroup.toErrorMessageGroup(IllegalStateException(info), protocol))
-                continue
-            }
+            val msgProtocols = messageGroup.allRawProtocols
+            val parentEventIds = messageGroup.allParentEventIds
+
+            try {
+                if (!protocols.checkAgainstProtocols(msgProtocols)) {
+                    logger.debug { "Messages with $msgProtocols protocols instead of $protocols are presented" }
+                    messageBatch.addGroups(messageGroup)
+                    continue
+                }
+
+                val decodedGroup = codec.decode(messageGroup)
 
-            messageGroup.runCatching(codec::decode).onSuccess { decodedGroup ->
                 if (decodedGroup.messagesCount < messageGroup.messagesCount) {
-                    parentEventId.onEvent("Decoded message group contains less messages (${decodedGroup.messagesCount}) than encoded one (${messageGroup.messagesCount})")
+                    parentEventIds.forEachEvent("Decoded message group contains less messages (${decodedGroup.messagesCount}) than encoded one (${messageGroup.messagesCount})")
                 }
 
                 messageBatch.addGroups(decodedGroup)
-            }.onFailure {
-                parentEventId.onErrorEvent("Failed to decode message group", messageGroup.messageIds, it)
-                messageBatch.addGroups(messageGroup.toErrorMessageGroup(it, protocol))
+            } catch (throwable: Throwable) {
+                val header = "Failed to decode message group"
+                val eventIds = parentEventIds.associateWith { messageEventId ->
+                    messageEventId.onErrorEvent(header, messageGroup.messageIds, throwable).id.let { errorEventId ->
+                        checkNotNull(EventUtils.toEventID(errorEventId)) {
```

Comment:
> checkNotNull

Redundant since `errorEventId` cannot be `null` and `EventUtils.toEventID` only returns `null` if `errorEventId` is `null`